---
title: "NGS analysis for gene regulation and epigenomics"
author: "Jacques Serizay, Cyril Matthey-Doret, Amaury Bignaud"
date: February 28, 2022"
output:
  ioslides_presentation: default
  slidy_presentation: default
---

# Introduction to Hi-C

Day 1


## Regulatory function of chromatin

* Chromatin made up of many DNA-bound proteins
* Dynamic interactions
* Chromatin state regulates gene expression

## A picture of the regulatory landscape

* Gene expression tells us what genes are affected during e.g. differentiation, disease...
* Epigenomics can tell us how / why they are activated

## Types of chromatin information

We will look at 3 scales of chromatin organization during this workshop:

<div class="columns-2">


![Adapted from [Illumina](https://www.illumina.com/content/dam/illumina-marketing/documents/applications/ngs-library-prep/for-all-you-seek-dna.pdf)](assets/chromatin_features.svg)

## Capturing chromosome conformation

* 3C methods capture the 3D structure of the genome
* Originally relied on qPCR, Hi-C is the NGS variant

```{r, out.width = "90%",  fig.cap= "Hi-C experimental process", echo=FALSE}
knitr::include_graphics("assets/hic.png")
```

## Interpreting chromosome contact maps

```{r, out.width = "90%",  fig.cap= "Contact frequency reflects spatial organization", echo=FALSE}
knitr::include_graphics("assets/hic_map_overview.png")
```

## Interpreting chromosome contact maps

<div class="columns-2">
  ![](assets/structures_to_patterns.png){width=50%}

  - Diagonal gradient due to polymer behavior
  - Various patterns correspond to 3D structures
</div>

## Derivative methods

3C-based methods (digestion + religation):

* MicroC: DNAse instead of restriction enzyme
* Promoter capture: Hi-C with capture probes on promoters

Others:

* GAM: Cryoslicing of nuclei followed by sequencing
* SPRITE: Serial dilution and barcoding


## From reads to contact maps

Many "end-to-end" pipelines available:

* **nf-core/hic**
* Hi-C pro
* FanC
* hicstuff
* Juicer
* ...

Important to understand individual steps (biases, custom analyses, ...)

## Hi-C processing: Overview

General steps common to all Hi-C pipelines

![](assets/hic_pipeline.svg){width=70%}


## Hi-C processing: Read alignment

* Standard short read mapping
* Separate single-end alignment for forward and reverse
* Longer reads relative to fragment length: More chimeric reads

![](assets/chimeric.svg){width=80%}

## Hi-C processing: Read alignment

* Solutions generally built in pipelines
  + iterative alignment
  + in-silico fastq digestion
* Mostly important with longer reads

![](assets/iteralign.svg){width=80%}

```python
truncate read
while not mapped:
  extend read
  mapped = align read
```

## Hi-C processing: Merging BAMs

<div class="columns-2">
  - Merge forward and reverse reads by name
  - Filter out ambiguous alignment (low mapQ)
  - 1 pair of reads = 1 contact
  - The [pairs](https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md) format is commonly used

  ![](assets/pipeline/hic_pipeline_merge_bams.svg){width=70%}
</div>


## Hi-C processing: Fragment attribution

* Assign contacts to discrete segments (bins)
* Usually regular intervals of e.g. 10kb

![](assets/pipeline/hic_pipeline_attribution.svg){width=90%}

## Hi-C processing: The Matrix

<div class="columns-2">
  -  Sum contact events for each combination of bins
  -  Most combinations have 0 contacts: "sparse data"
  -  Storing NxN matrix with mostly 0 is impractical
    + Use sparse format instead

  ![](assets/pipeline/hic_pipeline_matrix.svg){width=60%}
</div>

![](assets/dense_sparse.svg){width=80%}

## Hi-C processing: Additional filters

Hi-C specific filters to remove uninformative events

> - Self-religating fragments
> - Undigested fragments
> - Duplicates filter

> ![](assets/filters.svg){width=100%}

## Hi-C processing: Balancing

Hi-C is susceptible to many biases affecting local coverage such as:

* GC content
* Chromatin accessibility
* Restriction site density

Matrix balancing is a normalization to reduce the impact of those biases.

![](assets/balancing.svg){width=60%}

## Hi-C processing: Balancing

Assumption: All bins (regions) have the same contact probability.

* Divide each pixel by row x col iteratively.
* Predefined number of iterations or until convergence.

```python
# Simplified ICE procedure
m, n = mat.shape
for _ in range(100):
  for x in range(m):
    row_mean = mat[x, :].mean()
    for y in range(n):
      col_mean = mat[:, y].mean()
      mat[x, y] = mat[x, y] / (row_mean * col_mean)
```

# Exercises

* Generate your own contact map from the reads, normalize and visualize it
