<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="J. Serizay">

<title>Developing R/Bioconductor packages for Genomics - 4&nbsp; Using S4 classes in a package</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../content/exercises/Day5_package-integration.html" rel="next">
<link href="../../content/exercises/Day3_Examples-test-data.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../content/exercises/Day1_Getting-started-with-Bioc.html">Exercises</a></li><li class="breadcrumb-item"><a href="../../content/exercises/Day4_S4-classes.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Using S4 classes in a package</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Developing R/Bioconductor packages for Genomics</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/js2264/Bioc-workshop/" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/program.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/prerequisites.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prerequesites</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/extra.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Extra resources</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Day 1</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture0_course-overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 0 - Introduction to the workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture1_Package_architecture.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 1 - R package architecture</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture2_Bioconductor_101.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 2 - Introduction to Bioconductor</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture3_Developer-toolkit.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 3 - R developer toolkit</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Day 2</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lecture4_Continuous_development-writing-documenting-testing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 4 - Continuous development</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lecture5_Bioconductor_201.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 5 - Bioconductor 201</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Day 3</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture6_Git_versionning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 6 - Git versioning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture7_Improving_package-data_vignettes-....html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 8 - Vignettes and data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Day 4</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture8_S4_classes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 7 - S4 classes in Bioconductor</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture9_Bioconductor_submission.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 9 - Bioconductor package submission</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Day 5</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day5/Lecture10_Other_package_types.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 10 - Other package types</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day5/Lecture11_Package_dissemination.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lecture 11 - Package dissemination</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Exercises</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/exercises/Day1_Getting-started-with-Bioc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Getting started with Bioconductor ecosystem</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/exercises/Day2_Developer-virtuous-circle.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Kickstarting an R package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/exercises/Day3_Examples-test-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Enriching package functionalities</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/exercises/Day4_S4-classes.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Using S4 classes in a package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/exercises/Day5_package-integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Improving package integration</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#framework" id="toc-framework" class="nav-link active" data-scroll-target="#framework"><span class="header-section-number">4.1</span> Framework</a></li>
  <li><a href="#preliminary-checks" id="toc-preliminary-checks" class="nav-link" data-scroll-target="#preliminary-checks"><span class="header-section-number">4.2</span> 1. Preliminary checks</a></li>
  <li><a href="#parsing-bigwig-files" id="toc-parsing-bigwig-files" class="nav-link" data-scroll-target="#parsing-bigwig-files"><span class="header-section-number">4.3</span> 2. Parsing bigwig files</a></li>
  <li><a href="#preparing-rowdata" id="toc-preparing-rowdata" class="nav-link" data-scroll-target="#preparing-rowdata"><span class="header-section-number">4.4</span> 3. Preparing rowData</a></li>
  <li><a href="#building-the-aggregatedcoverage-class" id="toc-building-the-aggregatedcoverage-class" class="nav-link" data-scroll-target="#building-the-aggregatedcoverage-class"><span class="header-section-number">4.5</span> 4. Building the <code>AggregatedCoverage</code> class</a></li>
  <li><a href="#plotting-aggregatedcoverage-objects" id="toc-plotting-aggregatedcoverage-objects" class="nav-link" data-scroll-target="#plotting-aggregatedcoverage-objects"><span class="header-section-number">4.6</span> 5. Plotting <code>AggregatedCoverage</code> objects</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info"><span class="header-section-number">4.7</span> Session Info</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/js2264/Bioc-workshop/edit/main/content/exercises/Day4_S4-classes.Rmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/js2264/Bioc-workshop/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Using S4 classes in a package</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Put together the set of functions from Day 2 into a single workflow;</li>
<li>Document a new S4 class and its constructor;</li>
<li>Create a method to handle plotting for your new S4 class</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>At any time, if you are lost or do not understand how functions in the</strong> <strong>proposed solution work, type <code>?&lt;function&gt;</code> in the R console </strong> <strong>and a help menu will appear.</strong></p>
<p>You can also check the help tab in the corresponding quadrant.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Reminder
</div>
</div>
<div class="callout-body-container callout-body">
<p>We aim to create a package which can plot the aggregated coverage of a genomic track over a set of <code>GRanges</code> of interest (at a fixed width).</p>
</div>
</div>
<section id="framework" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="framework"><span class="header-section-number">4.1</span> Framework</h2>
<p>Functions developed in Day 2 work well for a specific couple of <code>.bigwig</code>/<code>.bed</code> files. However, the more used to Bioconductor you get, the more you will realize that you might already have your genomic features of interest (TSSs, genebodies, regulatory elements, binding motifs, …) already imported in R.</p>
<p>To make this package more “usable” by a broader Bioconductor audience, we will move to a more generic framework. We will create a single S4 class named <code>AggregatedCoverage</code>. The <code>AggregatedCoverage</code> constructor function will take 2 inputs:</p>
<ol type="1">
<li>a named list of <code>.RleList</code> objects (<code>RleList</code> are <code>bigwig</code> files imported in <code>R</code>);</li>
<li>a <code>GRanges</code> object (containing genomic features of interest)</li>
</ol>
<p>For each <code>.bigwig</code> file, the coverage over <code>GRanges</code> will be extracted and the mean +/- conf. interval (CI) coverage scores will be computed, similarly to what has been done in Day 2.</p>
<p>The <code>AggregatedCoverage</code> class will be a direct adaptation of the <code>SummarizedExperiment</code> class. It will contain a <code>colData</code> (refering to the “samples”, i.e.&nbsp;the bigwig files), a <code>rowData</code> (describing the genomic distance to the center of the <code>GRanges</code> of interest) and exactly 3 assays: <code>mean</code>, <code>lowCI</code> and <code>highCI</code>.</p>
</section>
<section id="preliminary-checks" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="preliminary-checks"><span class="header-section-number">4.2</span> 1. Preliminary checks</h2>
<p>To simplify the development of this package, we will require that the provided <code>GRanges</code> are all the same width.</p>
<blockquote class="blockquote">
<p>Write a function that checks whether all the entries in a <code>GRanges</code> are of the same width and if so, return that width.</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>check_widths <span class="ot">&lt;-</span> <span class="cf">function</span>(granges) {</span>
<span id="cb1-2"><a href="#cb1-2"></a>    <span class="cf">if</span> (<span class="fu">length</span>(<span class="fu">unique</span>(GenomicRanges<span class="sc">::</span><span class="fu">width</span>(granges))) <span class="sc">!=</span> <span class="dv">1</span>) {</span>
<span id="cb1-3"><a href="#cb1-3"></a>        <span class="fu">stop</span>(<span class="st">"The provided GRanges should all have the same width."</span>)</span>
<span id="cb1-4"><a href="#cb1-4"></a>    }</span>
<span id="cb1-5"><a href="#cb1-5"></a>    <span class="cf">else</span> {</span>
<span id="cb1-6"><a href="#cb1-6"></a>        <span class="fu">return</span>(GenomicRanges<span class="sc">::</span><span class="fu">width</span>(granges[<span class="dv">1</span>]))</span>
<span id="cb1-7"><a href="#cb1-7"></a>    }</span>
<span id="cb1-8"><a href="#cb1-8"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
<blockquote class="blockquote">
<p>Create a <code>checks.R</code> file in <code>R/</code> and add your check function to it. Don’t forget to document it!!</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>usethis<span class="sc">::</span><span class="fu">use_r</span>(<span class="st">'checks'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
<blockquote class="blockquote">
<p>Do you need to export this function?</p>
</blockquote>
</section>
<section id="parsing-bigwig-files" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="parsing-bigwig-files"><span class="header-section-number">4.3</span> 2. Parsing bigwig files</h2>
<p>Another “module” required in this package is a function to 1) import a bigwig file in memory, 2) subset it to a provided <code>GRanges</code> and 3) compute mean and confidence interval values. The output of this function should be a data.frame, with several columns: the bp position from the center of the input <code>GRanges</code> and the mean, lower CI and higher CI values of coverage of the input <code>.bigwig</code>.</p>
<blockquote class="blockquote">
<p>Write a <code>parseBigWig</code> function to do this.</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>parseBigWig <span class="ot">&lt;-</span> <span class="cf">function</span>(file, granges) {</span>
<span id="cb3-2"><a href="#cb3-2"></a>    rtracklayer<span class="sc">::</span><span class="fu">import</span>(file, <span class="at">as =</span> <span class="st">"Rle"</span>)[granges] <span class="sc">|&gt;</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>        <span class="fu">as.matrix</span>() <span class="sc">|&gt;</span></span>
<span id="cb3-4"><a href="#cb3-4"></a>        <span class="fu">t</span>() <span class="sc">|&gt;</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>        <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb3-6"><a href="#cb3-6"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">pos =</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="fu">unique</span>(<span class="fu">width</span>(sub_granges))<span class="sc">/</span><span class="dv">2</span>, <span class="fu">unique</span>(<span class="fu">width</span>(sub_granges))<span class="sc">/</span><span class="dv">2-1</span>, <span class="at">by =</span> <span class="dv">1</span>)) <span class="sc">|&gt;</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>        tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="sc">-</span>pos, <span class="at">names_to =</span> <span class="st">'idx'</span>, <span class="at">values_to =</span> <span class="st">'cov'</span>) <span class="sc">|&gt;</span></span>
<span id="cb3-8"><a href="#cb3-8"></a>        dplyr<span class="sc">::</span><span class="fu">group_by</span>(pos) <span class="sc">|&gt;</span></span>
<span id="cb3-9"><a href="#cb3-9"></a>        dplyr<span class="sc">::</span><span class="fu">summarize</span>(</span>
<span id="cb3-10"><a href="#cb3-10"></a>            <span class="at">score =</span> <span class="fu">mean</span>(cov, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb3-11"><a href="#cb3-11"></a>            <span class="at">median =</span> <span class="fu">median</span>(cov, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb3-12"><a href="#cb3-12"></a>            <span class="at">sd =</span> <span class="fu">sd</span>(cov, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb3-13"><a href="#cb3-13"></a>            <span class="at">count =</span> dplyr<span class="sc">::</span><span class="fu">n</span>()</span>
<span id="cb3-14"><a href="#cb3-14"></a>        ) <span class="sc">|&gt;</span></span>
<span id="cb3-15"><a href="#cb3-15"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb3-16"><a href="#cb3-16"></a>            <span class="at">se =</span> sd<span class="sc">/</span><span class="fu">sqrt</span>(count),</span>
<span id="cb3-17"><a href="#cb3-17"></a>            <span class="at">ci_low =</span> score <span class="sc">-</span> <span class="fu">qt</span>(<span class="dv">1</span> <span class="sc">-</span> (<span class="fl">0.05</span> <span class="sc">/</span> <span class="dv">2</span>), count <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">*</span> se,</span>
<span id="cb3-18"><a href="#cb3-18"></a>            <span class="at">ci_high =</span> score <span class="sc">+</span> <span class="fu">qt</span>(<span class="dv">1</span> <span class="sc">-</span> (<span class="fl">0.05</span> <span class="sc">/</span> <span class="dv">2</span>), count <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">*</span> se</span>
<span id="cb3-19"><a href="#cb3-19"></a>        )</span>
<span id="cb3-20"><a href="#cb3-20"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
<blockquote class="blockquote">
<p>Create a <code>parse.R</code> file in <code>R/</code> and add your parsing function to it. Don’t forget to document it!!</p>
</blockquote>
</section>
<section id="preparing-rowdata" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="preparing-rowdata"><span class="header-section-number">4.4</span> 3. Preparing rowData</h2>
<p>Since the <code>AggregatedCoverage</code> class is adapted from the <code>SummarizedExperiment</code> class, the information regarding the genomic distance to the center of the <code>GRanges</code> of interest can be stored in a <code>rowData</code> object.</p>
<blockquote class="blockquote">
<p>Write a simple <code>prepareRowData</code> function which takes a <code>width</code> argument (the width of the <code>GRanges</code>) and returns a data.frame which contains a single <code>pos</code> column of that length, centered at 0 (e.g.&nbsp;for <code>width</code> == 2000, the <code>pos</code> column would be a sequence from -1000 to 999 (length of 2000)).</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>prepareRowData <span class="ot">&lt;-</span> <span class="cf">function</span>(width) {</span>
<span id="cb4-2"><a href="#cb4-2"></a>    <span class="fu">data.frame</span>(</span>
<span id="cb4-3"><a href="#cb4-3"></a>        <span class="at">pos =</span> <span class="fu">seq</span>(<span class="sc">-</span>width_gr<span class="sc">/</span><span class="dv">2</span>, width_gr<span class="sc">/</span><span class="dv">2-1</span>, <span class="at">by =</span> <span class="dv">1</span>)</span>
<span id="cb4-4"><a href="#cb4-4"></a>    )</span>
<span id="cb4-5"><a href="#cb4-5"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
<blockquote class="blockquote">
<p>Add this function to a <code>utils.R</code> file.</p>
</blockquote>
</section>
<section id="building-the-aggregatedcoverage-class" class="level2" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="building-the-aggregatedcoverage-class"><span class="header-section-number">4.5</span> 4. Building the <code>AggregatedCoverage</code> class</h2>
<p>Everything is now ready to create an <code>AggregatedCoverage</code> object.</p>
<blockquote class="blockquote">
<p>Fill out the <code>...</code> fields in this template to create an <code>AggregatedCoverage</code> object, using a <code>granges</code> and a colData containig a list of <code>.bigwig</code> files as inputs (see below).</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>granges <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">'../../../Share/Day3/Scc1-peaks.narrowPeak'</span>))</span>
<span id="cb5-2"><a href="#cb5-2"></a>granges <span class="ot">&lt;-</span> <span class="fu">resize</span>(granges, <span class="at">width =</span> <span class="dv">1</span>, <span class="at">fix =</span> <span class="st">'center'</span>) <span class="sc">|&gt;</span> <span class="fu">resize</span>(<span class="at">width =</span> <span class="dv">5000</span>, <span class="at">fix =</span> <span class="st">'center'</span>)</span>
<span id="cb5-3"><a href="#cb5-3"></a>colData <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb5-4"><a href="#cb5-4"></a>    <span class="at">file =</span> <span class="fu">c</span>(</span>
<span id="cb5-5"><a href="#cb5-5"></a>        here<span class="sc">::</span><span class="fu">here</span>(<span class="st">'../../../Share/Day4/Scc1-calibrated_rep1.bw'</span>),</span>
<span id="cb5-6"><a href="#cb5-6"></a>        here<span class="sc">::</span><span class="fu">here</span>(<span class="st">'../../../Share/Day4/Scc1-calibrated_rep2.bw'</span>)</span>
<span id="cb5-7"><a href="#cb5-7"></a>    )</span>
<span id="cb5-8"><a href="#cb5-8"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># 1. Check width of input GRanges</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>width_gr <span class="ot">&lt;-</span> <span class="fu">check_widths</span>(...)</span>
<span id="cb6-3"><a href="#cb6-3"></a></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co"># 2. For each bigwig: import it and compute stats over granges</span></span>
<span id="cb6-5"><a href="#cb6-5"></a>l_cvgs <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_len</span>(<span class="fu">nrow</span>(colData)), <span class="cf">function</span>(K){</span>
<span id="cb6-6"><a href="#cb6-6"></a>    <span class="fu">message</span>(<span class="st">"Parsing sample "</span>, K, <span class="st">"/"</span>, <span class="fu">nrow</span>(colData))</span>
<span id="cb6-7"><a href="#cb6-7"></a>    file <span class="ot">&lt;-</span> colData[K, <span class="st">'...'</span>]</span>
<span id="cb6-8"><a href="#cb6-8"></a>    <span class="fu">parseBigWig</span>(file, ...)</span>
<span id="cb6-9"><a href="#cb6-9"></a>})</span>
<span id="cb6-10"><a href="#cb6-10"></a></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="co"># 2. Prepare rowData</span></span>
<span id="cb6-12"><a href="#cb6-12"></a>rowData <span class="ot">&lt;-</span> <span class="fu">prepareRowData</span>(...)</span>
<span id="cb6-13"><a href="#cb6-13"></a></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co"># 3. Create an `AggregatedCoverage` object</span></span>
<span id="cb6-15"><a href="#cb6-15"></a>AC <span class="ot">&lt;-</span> methods<span class="sc">::</span><span class="fu">new</span>(</span>
<span id="cb6-16"><a href="#cb6-16"></a>    <span class="st">"..."</span>,</span>
<span id="cb6-17"><a href="#cb6-17"></a>    SummarizedExperiment<span class="sc">::</span><span class="fu">SummarizedExperiment</span>(</span>
<span id="cb6-18"><a href="#cb6-18"></a>        <span class="at">rowData =</span> ...,</span>
<span id="cb6-19"><a href="#cb6-19"></a>        <span class="at">colData =</span> ...,</span>
<span id="cb6-20"><a href="#cb6-20"></a>        <span class="at">assays =</span> <span class="fu">list</span>(</span>
<span id="cb6-21"><a href="#cb6-21"></a>            <span class="st">'mean'</span> <span class="ot">=</span> <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(<span class="at">... =</span> , <span class="st">'[['</span>, <span class="st">'...'</span>)),</span>
<span id="cb6-22"><a href="#cb6-22"></a>            <span class="st">'lowCI'</span> <span class="ot">=</span> <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(..., <span class="st">'[['</span>, <span class="st">'...'</span>)),</span>
<span id="cb6-23"><a href="#cb6-23"></a>            <span class="st">'highCI'</span> <span class="ot">=</span> <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(..., <span class="st">'[['</span>, <span class="st">'...'</span>))</span>
<span id="cb6-24"><a href="#cb6-24"></a>        )</span>
<span id="cb6-25"><a href="#cb6-25"></a>    )</span>
<span id="cb6-26"><a href="#cb6-26"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
<p>Once you have it working, you need to turn it into a constructor! But first, you need to properly define the S4 class you’re creating in this package.</p>
<blockquote class="blockquote">
<p>Create a <code>AllClass.R</code> file and set a new method in this file.</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>methods<span class="sc">::</span><span class="fu">setClass</span>(<span class="st">"AggregatedCoverage"</span>, <span class="at">contains =</span> <span class="fu">c</span>(<span class="st">"SummarizedExperiment"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
<blockquote class="blockquote">
<p>In the same file, create a constructor function, wrapping the previous chunk in which you created a single <code>AggregatedCoverage</code> object. By convention, the constructor function bears the same name than the class itself (<code>AggregatedCoverage</code>).</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>AggregatedCoverage <span class="ot">&lt;-</span> <span class="cf">function</span>(granges, colData) {</span>
<span id="cb8-2"><a href="#cb8-2"></a></span>
<span id="cb8-3"><a href="#cb8-3"></a>    <span class="co"># Check width of input GRanges</span></span>
<span id="cb8-4"><a href="#cb8-4"></a>    width_gr <span class="ot">&lt;-</span> <span class="fu">check_widths</span>(granges)</span>
<span id="cb8-5"><a href="#cb8-5"></a></span>
<span id="cb8-6"><a href="#cb8-6"></a>    <span class="co"># Import each bigwig and compute stats</span></span>
<span id="cb8-7"><a href="#cb8-7"></a>    l_cvgs <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_len</span>(<span class="fu">nrow</span>(colData)), <span class="cf">function</span>(K){</span>
<span id="cb8-8"><a href="#cb8-8"></a>        <span class="fu">message</span>(<span class="st">"Parsing sample "</span>, K, <span class="st">"/"</span>, <span class="fu">nrow</span>(colData))</span>
<span id="cb8-9"><a href="#cb8-9"></a>        file <span class="ot">&lt;-</span> colData[K, <span class="st">'file'</span>]</span>
<span id="cb8-10"><a href="#cb8-10"></a>        <span class="fu">parseBigWig</span>(file, rowRanges)</span>
<span id="cb8-11"><a href="#cb8-11"></a>    })</span>
<span id="cb8-12"><a href="#cb8-12"></a></span>
<span id="cb8-13"><a href="#cb8-13"></a>    <span class="co"># Prepare rowData</span></span>
<span id="cb8-14"><a href="#cb8-14"></a>    rowData <span class="ot">&lt;-</span> <span class="fu">prepareRowData</span>(width_gr)</span>
<span id="cb8-15"><a href="#cb8-15"></a></span>
<span id="cb8-16"><a href="#cb8-16"></a>    <span class="co"># Create an `AggregatedCoverage` object</span></span>
<span id="cb8-17"><a href="#cb8-17"></a>    AC <span class="ot">&lt;-</span> methods<span class="sc">::</span><span class="fu">new</span>(</span>
<span id="cb8-18"><a href="#cb8-18"></a>        <span class="st">"AggregatedCoverage"</span>,</span>
<span id="cb8-19"><a href="#cb8-19"></a>        SummarizedExperiment<span class="sc">::</span><span class="fu">SummarizedExperiment</span>(</span>
<span id="cb8-20"><a href="#cb8-20"></a>            <span class="at">rowData =</span> rowData,</span>
<span id="cb8-21"><a href="#cb8-21"></a>            <span class="at">colData =</span> colData,</span>
<span id="cb8-22"><a href="#cb8-22"></a>            <span class="at">assays =</span> <span class="fu">list</span>(</span>
<span id="cb8-23"><a href="#cb8-23"></a>                <span class="st">'mean'</span> <span class="ot">=</span> <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(l_cvgs, <span class="st">'[['</span>, <span class="st">'score'</span>)),</span>
<span id="cb8-24"><a href="#cb8-24"></a>                <span class="st">'lowCI'</span> <span class="ot">=</span> <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(l_cvgs, <span class="st">'[['</span>, <span class="st">'ci_low'</span>)),</span>
<span id="cb8-25"><a href="#cb8-25"></a>                <span class="st">'highCI'</span> <span class="ot">=</span> <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(l_cvgs, <span class="st">'[['</span>, <span class="st">'ci_high'</span>))</span>
<span id="cb8-26"><a href="#cb8-26"></a>            )</span>
<span id="cb8-27"><a href="#cb8-27"></a>        )</span>
<span id="cb8-28"><a href="#cb8-28"></a>    )</span>
<span id="cb8-29"><a href="#cb8-29"></a>    <span class="fu">return</span>(AC)</span>
<span id="cb8-30"><a href="#cb8-30"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
<blockquote class="blockquote">
<p>Now add documentation to this new file. Which function should you export?</p>
</blockquote>
<blockquote class="blockquote">
<p>Try your function.</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>granges <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">'../../../Share/Day3/Scc1-peaks.narrowPeak'</span>))</span>
<span id="cb9-2"><a href="#cb9-2"></a>granges <span class="ot">&lt;-</span> <span class="fu">resize</span>(granges, <span class="at">width =</span> <span class="dv">1</span>, <span class="at">fix =</span> <span class="st">'center'</span>) <span class="sc">|&gt;</span> <span class="fu">resize</span>(<span class="at">width =</span> <span class="dv">5000</span>, <span class="at">fix =</span> <span class="st">'center'</span>)</span>
<span id="cb9-3"><a href="#cb9-3"></a>colData <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb9-4"><a href="#cb9-4"></a>    <span class="at">file =</span> <span class="fu">c</span>(</span>
<span id="cb9-5"><a href="#cb9-5"></a>        here<span class="sc">::</span><span class="fu">here</span>(<span class="st">'../../../Share/Day3/Scc1-calibrated_rep1.bw'</span>),</span>
<span id="cb9-6"><a href="#cb9-6"></a>        here<span class="sc">::</span><span class="fu">here</span>(<span class="st">'../../../Share/Day3/Scc1-calibrated_rep2.bw'</span>)</span>
<span id="cb9-7"><a href="#cb9-7"></a>    )</span>
<span id="cb9-8"><a href="#cb9-8"></a>)</span>
<span id="cb9-9"><a href="#cb9-9"></a>AC <span class="ot">&lt;-</span> <span class="fu">AggregatedCoverage</span>(granges, colData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
<blockquote class="blockquote">
<p>Does it work? What do you need to adapt? Change the <code>parseBigWig</code> function to only retain <code>GRanges</code> contained within the input coverage track.</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>parseBigWig <span class="ot">&lt;-</span> <span class="cf">function</span>(file, granges) {</span>
<span id="cb10-2"><a href="#cb10-2"></a>    cov <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(file, <span class="at">as =</span> <span class="st">"Rle"</span>)</span>
<span id="cb10-3"><a href="#cb10-3"></a>    l <span class="ot">&lt;-</span> <span class="fu">filterGRanges</span>(<span class="fu">list</span>(<span class="at">coverage =</span> cov, <span class="at">features =</span> granges))</span>
<span id="cb10-4"><a href="#cb10-4"></a>    sub_granges <span class="ot">&lt;-</span> l[[<span class="st">'features'</span>]]</span>
<span id="cb10-5"><a href="#cb10-5"></a>    cov[sub_granges] <span class="sc">|&gt;</span></span>
<span id="cb10-6"><a href="#cb10-6"></a>        <span class="fu">as.matrix</span>() <span class="sc">|&gt;</span></span>
<span id="cb10-7"><a href="#cb10-7"></a>        <span class="fu">t</span>() <span class="sc">|&gt;</span></span>
<span id="cb10-8"><a href="#cb10-8"></a>        ...</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
</section>
<section id="plotting-aggregatedcoverage-objects" class="level2" data-number="4.6">
<h2 data-number="4.6" class="anchored" data-anchor-id="plotting-aggregatedcoverage-objects"><span class="header-section-number">4.6</span> 5. Plotting <code>AggregatedCoverage</code> objects</h2>
<p>Now that we have a working S4 class, we can think about setting a new <em>method</em> to plot these scores. The plot method can be adapted from that developed in Day 2, but will not be a dedicated <strong>function</strong>, but a <strong>method</strong> associated with the <code>plot</code> generic function.</p>
<blockquote class="blockquote">
<p>In the <code>plot.R</code> file, set a new <code>plot</code> method (<code>setMethod</code>) for the <code>AggregatedCoverage</code> signature.</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="fu">setMethod</span>(<span class="st">"plot"</span>, <span class="st">"AggregatedCoverage"</span>, <span class="cf">function</span>(x) {</span>
<span id="cb11-2"><a href="#cb11-2"></a>    df <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_len</span>(<span class="fu">ncol</span>(x)), <span class="cf">function</span>(K) {</span>
<span id="cb11-3"><a href="#cb11-3"></a>        <span class="fu">data.frame</span>(</span>
<span id="cb11-4"><a href="#cb11-4"></a>            <span class="at">file =</span> <span class="fu">colData</span>(x)<span class="sc">$</span>file[K],</span>
<span id="cb11-5"><a href="#cb11-5"></a>            <span class="at">K =</span> K,</span>
<span id="cb11-6"><a href="#cb11-6"></a>            <span class="at">pos =</span> <span class="fu">rowData</span>(x)<span class="sc">$</span>pos,</span>
<span id="cb11-7"><a href="#cb11-7"></a>            <span class="at">mean =</span> <span class="fu">assay</span>(x, <span class="st">'mean'</span>)[, K],</span>
<span id="cb11-8"><a href="#cb11-8"></a>            <span class="at">lowCI =</span> <span class="fu">assay</span>(x, <span class="st">'lowCI'</span>)[, K],</span>
<span id="cb11-9"><a href="#cb11-9"></a>            <span class="at">highCI =</span> <span class="fu">assay</span>(x, <span class="st">'highCI'</span>)[, K]</span>
<span id="cb11-10"><a href="#cb11-10"></a>        )</span>
<span id="cb11-11"><a href="#cb11-11"></a>    }) <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">bind_rows</span>()</span>
<span id="cb11-12"><a href="#cb11-12"></a>    <span class="fu">ggplot</span>(df, <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb11-13"><a href="#cb11-13"></a>        <span class="at">x =</span> pos, <span class="at">y =</span> mean, <span class="at">ymin =</span> lowCI, <span class="at">ymax =</span> highCI,</span>
<span id="cb11-14"><a href="#cb11-14"></a>        <span class="at">col =</span> <span class="fu">basename</span>(file), <span class="at">fill =</span> <span class="fu">basename</span>(file)</span>
<span id="cb11-15"><a href="#cb11-15"></a>    )) <span class="sc">+</span></span>
<span id="cb11-16"><a href="#cb11-16"></a>        <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb11-17"><a href="#cb11-17"></a>        <span class="fu">geom_ribbon</span>(<span class="at">col =</span> <span class="cn">NA</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>)</span>
<span id="cb11-18"><a href="#cb11-18"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
<blockquote class="blockquote">
<p>Try plotting your <code>AC</code> object.</p>
</blockquote>
<details>
<summary style="color: #ff7f00; font-weight: bold">
Show code
</summary>
<p>
</p><div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="fu">plot</span>(AC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p></p>
</details>
<p><br></p>
</section>
<section id="session-info" class="level2" data-number="4.7">
<h2 data-number="4.7" class="anchored" data-anchor-id="session-info"><span class="header-section-number">4.7</span> Session Info</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] compiler_4.3.2  fastmap_1.1.1   cli_3.6.1       tools_4.3.2    
 [5] htmltools_0.5.7 rmarkdown_2.25  knitr_1.45      jsonlite_1.8.7 
 [9] xfun_0.41       digest_0.6.33   rlang_1.1.2     evaluate_0.23  </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../content/exercises/Day3_Examples-test-data.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Enriching package functionalities</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../content/exercises/Day5_package-integration.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Improving package integration</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">Developing R/Bioconductor packages for Genomics |<br>
J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>



</body></html>