# Improving package integration

```{r echo = FALSE, warning = FALSE, message = FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

- **Overarching goal:** Improve package embedding in Bioconductor ecosystem
- **Specific aims:**

    - Add suppporting functions to access AnnotationHub-hosted files
    - Create a pkgdown website for package documentation

## 1. Improve examples using TxDb and AnnotationHub resources

`TxDb` packages provide transcript annotations for model organisms. 

> Find the `TxDb` package associated with UCSC-provided human gene annotations and 
install it. **You will need annotations from the `hg19` human genome reference!**

```{r}
BiocManager::install('TxDb.Hsapiens.UCSC.hg19.knownGene')
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene
```

`GenomicFeatures` is a package designed in order to facilitate conversion of 
`TxDb` objects into `GRanges` objects. 

> Which function from `GenomicFeatures` can be used to find the gene loci for all genes present in the human `TxDb`? 

```{r}
GenomicFeatures::genes(txdb)
```

AnnotationHub provides extensive human NGS datasets. 

> Can you query the AnnotationHub to explore which datasets are available? 
Which species are most represented? 

```{r}
ah <- AnnotationHub::AnnotationHub()
ah_df <- AnnotationHub::mcols(ah)
table(ah_df$species) |> sort() |> tail(10)
table(ah_df$sourcetype) |> sort() |> tail(10)
```

The Epigenome roadmap project by the Broad Institute performed H3K4me3 ChIP-seq
in human. They generated a library labelled 'LL227' ([GEO ID: GSM409308](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM409308)).

> Can you query the AnnotationHub for this library?

```{r}
AnnotationHub::query(ah, 'LL227')
```

> Which signal corresponds to a ChIP-seq coverage track? How can you download the associated file? 

```{r}
bw <- ah[['AH34904']]
```

> What is the path of the file now stored on disk? How can you load the file in R? 

```{r}
BiocIO::resource(bw)
cov <- BiocIO::import(bw)
```

> Create a `colData` data frame containing a single `file` column containing this file,
then compute an `AggregatedCoverage` object over human TSSs (+/- 5kb). 

```{r}
genes <- GenomicFeatures::genes(txdb)
tss <- GenomicRanges::resize(genes, width = 1, fix = 'start')
tss_extended <- GenomicRanges::resize(tss, width = 10000, fix = 'center')
colData <- data.frame(file = BiocIO::resource(bw))
AC <- AggregatedCoverage(tss_extended, colData)
AC
```

> Can you plot the results? 

```{r}
plot(AC)
```

> Create a new vignette to illustrate use of AnnotationHub data with your package.

```{r}

```


## 2. Create pkgdown website and host it on GitHub

> Check `pkgdown` documentation. What is the recommended way to add pkgdown support 
for your package? 

```{r}
usethis::use_pkgdown()
```

> Can you now build your package? What does the building step do exactly?

```{r}
pkgdown::build_site()
```

> Comment on why setting up a pkgdown website helps with checking that your package reliably works? 

> Commit your changes. Did that include your newly created `docs` folder? Why? 

> Remove `docs` from `.gitignore`, then commit the `docs/` folder with git. 

> Now push all your commits to your github remote repository, navigate to its webpage, and enable `Pages` deployment from the `docs` folder. Check the resulting website. 
