---
title: "Day 1: Getting started with Bioconductor ecosystem"
output: 
    rmdformats::readthedown: 
        highlight: tango
        preserve_yaml: true
        df_print: tibble
        toc_depth: 4
        css: ../../custom.css
---

```{r echo = FALSE, warning = FALSE, message = FALSE}
knitr::opts_chunk$set(eval = FALSE)
library(Biostrings)
library(GenomicRanges)
library(rtracklayer)
library(ggplot2)
```

- **Overarching goal:** Getting started with Bioconductor ecosystem
- **Specific aims:** 

    - Importing files from various formats in R
    - Handling `DNAStringSet`
    - Handling `GRanges`
    - Finding a specific sequence within larger ones
    - Computing distances between sets of `GRanges`

**At any time, if you are lost or do not understand how functions in the**
**proposed solution work, type `?<function>` in the R console **
**and a help menu will appear.**

## 1. Importing ce11 genome sequence in R

ce11 genome sequence is stored in a `.fa` file in `Share/Day1/`.  

> Find the appropriate Bioconductor's function to import fasta files in R and 
use it to import ce11 genome sequence.

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
library(Biostrings)
gseq <- readDNAStringSet(here::here('Share/Day1/ce11.fa'))
```
</p></details><br>

> What are the chromosome names used in the fasta file? 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
gseq
names(gseq)
```
</p></details><br>

> What is the difference between the `length` and the `width` of a `DNAStringSet` object?

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
width(gseq)
length(gseq)
```
</p></details><br>

## 2. Importing ce11 gene annotations in R

ce11 gene annotations are stored in a `gtf` file in `Share/Day1/`.  

> Find the appropriate Bioconductor's function to import gtf files in R and 
use it to import ce11 gene annotations.

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
library(rtracklayer)
genes <- import(here::here('Share/Day1/ce11_annotations.gtf'))
```
</p></details><br>

> How many genes are they? What's the median gene length? Which tissues are genes expressed in?

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
length(genes)
median(width(genes))
table(genes$tissue)
```
</p></details><br>

> What are the chromosome names used in this gene annotations file? How many genes map to each chromosome? 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
seqlevels(genes)
table(seqnames(genes))
```
</p></details><br>

> Why are the chromosome names (`seqlevels`) ordered this way? Can you reorder them? 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
seqlevels(genes) <- seqlevels(genes)[c(5, 4, 6, 3, 1, 2, 7)]
```
</p></details><br>

## 3. Wrangling gene annotations 

> Is mitochondrial chromosome present in the genome sequence?

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
seqlevels(gseq)
```
</p></details><br>

> Try to prune genes to remove all those mapping over mitochondrial chromosome. 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
genes[seqnames(genes) != 'MtDNA']
seqlevels(genes, pruning.mode = 'coarse') <- seqlevels(genes)[1:6]
```
</p></details><br>

> Compare the chromosome naming nomenclature for genes and chromosome sequences. Are they the same? Can you make them match? 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
seqlevelsStyle(gseq)
seqlevelsStyle(genes)
seqlevelsStyle(genes) <- "UCSC"
```
</p></details><br>

## 4. Extract tissue-specific TSSs

Now that we have two concordant objects, we would like to focus on sets of 
tissue-specific TSSs.

> Re-focus the genes to center them at their TSSs. Choose a window size that you find appropriate (hint: we'll late focus on TATA box, so the window has to at least encompass it!)

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
WINDOW_SIZE <- 300
TSSs <- resize(genes, fix = 'start', width = 1) |> resize(fix = 'center', width = WINDOW_SIZE)
```
</p></details><br>

To iterate over a set of values (e.g. a set of tissues...), 
one should favor `lapply` function rather than `for` loops. 
`lapply` outputs a list of elements, one for each element in the vector used 
in the input. Check `?lapply` for more information if you are not familiar 
with it. 

> Make a list of 6 elements, containing **forward** TSSs specific of each of 
the 5 main tissues, and the ubiquitous TSSs. 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
tissues <- c("Germline", "Neurons", "Muscle", "Hypod.", "Intest.", "Ubiq.")
TSS_list <- lapply(tissues, function(tissue) {
    TSSs[TSSs$tissue == tissue & strand(TSSs) == '+']
})
names(TSS_list) <- tissues
lengths(TSS_list)
```
</p></details><br>

## 5. Find TATA boxes near intestine-specific TSSs

> Start by subseting full genome sequence to only the sequences at intestinal TSSs. 
Check `?DNAStringSet` to see how to subset a `DNAStringSet`.

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
intest_TSSs <- TSS_list[['Intest.']]
intest_TSS_seqs <- gseq[intest_TSSs]
intest_TSS_seqs
table(width(intest_TSS_seqs))
```
</p></details><br>

> Read `?matchPattern` for information on how to find a given sequence in a `DNAStringSet`. What is the difference between `matchPattern()` and `vmatchPattern()`? 

> Is there a TATA box ("TATAAA") in the first intestine TSS sequence?

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
pattern <- 'TATAAA'
seq <- intest_TSS_seqs[[1]]
matchPattern(pattern, seq)
```
</p></details><br>

> Map TATA boxes across all intestine TSS sequences. 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
pattern <- 'TATAAA'
TATA_boxes <- vmatchPattern(pattern, intest_TSS_seqs)
TATA_boxes
```
</p></details><br>

> How to parse `MIndex` objects? Check `?MIndex` if in doubt...

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
positions <- unlist(startIndex(TATA_boxes))
head(positions)
```
</p></details><br>

> Plot the distance between TATA box and intestine-specific TSSs.

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
library(ggplot2)
df <- data.frame(pos = positions - WINDOW_SIZE/2)
ggplot(df, aes(x = pos)) + 
    geom_histogram(binwidth = 10) + 
    labs(x = "Distance to TSS", y = "# of motifs")
```
</p></details><br>

## 7. Function wrapping

We now know how to 1) parse `GRanges`, 2) map a chosen motif ("TATAAA") over a set of sequences and 3)
plot the distance between this motif and the center of sequences. 

> Create a function which takes a set of sequences and a chosen motif as input, and returns a plot of the distance between 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
plotMotifDistance <- function(seqs, motif) {
    motif_occurences <- vmatchPattern(motif, seqs)
    positions <- unlist(startIndex(motif_occurences))
    df <- data.frame(pos = positions - WINDOW_SIZE/2)
    ggplot(df, aes(x = pos)) + 
        geom_histogram(binwidth = 10) + 
        labs(
            x = "Distance to TSS", 
            y = glue::glue("# of {motif} motifs"), 
            caption = glue::glue("{length(positions)} motifs found amongst {length(seqs)} sequences")
        )
}
plotMotifDistance(intest_TSS_seqs, "TATAAA")
```
</p></details><br>

> Add an argument to precise number of possible mismatches when looking for the motif. Also adapt the `vmatchPattern()` function!

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
plotMotifDistance <- function(seqs, motif, n_mismatch = 1) {
    motif_occurences <- vmatchPattern(motif, seqs, max.mismatch = n_mismatch)
    positions <- unlist(startIndex(motif_occurences))
    df <- data.frame(pos = positions - WINDOW_SIZE/2)
    ggplot(df, aes(x = pos)) + 
        geom_histogram(binwidth = 10) + 
        labs(
            x = "Distance to TSS", 
            y = glue::glue("# of {motif} motifs"), 
            caption = glue::glue("{length(positions)} motifs found amongst {length(seqs)} sequences")
        )
}
plotMotifDistance(intest_TSS_seqs, "TATAAA", n_mismatch = 0)
```
</p></details><br>

> The real TATA box consensus sequence is closer to "TATAWAA". Can you further adapt `vmatchPattern()` function so it can accept all IUPAC code (e.g. W = A/T).

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
plotMotifDistance <- function(seqs, motif, n_mismatch = 1) {
    motif_occurences <- vmatchPattern(motif, seqs, max.mismatch = n_mismatch, fixed = FALSE)
    positions <- unlist(startIndex(motif_occurences))
    df <- data.frame(pos = positions - WINDOW_SIZE/2)
    ggplot(df, aes(x = pos)) + 
        geom_histogram(binwidth = 5) + 
        labs(
            x = "Distance to TSS", 
            y = glue::glue("# of {motif} motifs"), 
            caption = glue::glue("{length(positions)} motifs found amongst {length(seqs)} sequences")
        )
}
plotMotifDistance(intest_TSS_seqs, "TATAWAA", n_mismatch = 0)
```
</p></details><br>

> Run this function to find TATA boxes in sequences from each set of TSSs.

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
lapply(names(TSS_list), function(tissue) {
    TSSs <- TSS_list[[tissue]]
    plotMotifDistance(
        gseq[TSSs], 
        "TATAWAA", 
        n_mismatch = 0
    ) + ggtitle(glue::glue("{tissue} TSSs"))
})
```
</p></details><br>

> Run this function to find PolII Initiator (Inr) motif ("YYANWYY") in sequences from each set of TSSs. Comment.

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
lapply(names(TSS_list), function(tissue) {
    TSSs <- TSS_list[[tissue]]
    plotMotifDistance(
        gseq[TSSs], 
        "YYANWYY", 
        n_mismatch = 1
    ) + ggtitle(glue::glue("{tissue} TSSs"))
})
```
</p></details><br>

> How to deal with minus-oriented `GRanges`? Propose a solution.

## Session Info

```{r}
sessionInfo()
```

